### PR TITLE
fix unreported issue where server dxf export for espg's with reverse axes doesn't work

### PR DESCRIPTION
**Pull request by tudorbarascu**
The dxf export didn't work for reverse axis projections (north, east)
as the BBOX for WMS 1.3.0 comes in South, West, North, East and the
dxf export treated it like West, South, East, North.

The logic is taken from the normal GetMap, GetPrint requests but somehow nobody reported it in for the DXF export.

Hope I did it well. Tested it and now it works. For some reason I cannot momentarily build with tests but would like to add some python tests to cover this later on.

Tested manually for EPSG 3844 and 3857 and 4326.

### CODE REVIEW

**Comment by tudorbarascu** [comment]
CI failure unrelated

**Comment by elpaso** [comment]
LGTM but it would be nice to have a test for this, so that it won't break in the future.

**Comment by tudorbarascu** [comment]
@elpaso Thanks for reviewing. I promise I'll make a test. Would love to make a test now but it would have to wait a few days/weeks so I can find several hours as I suspect it will take that much and I'm really short on time. I'll try to squeeze the work in  asap.

**Comment by elpaso** [comment]
> @elpaso Thanks for reviewing. I promise I'll make a test. Would love to make a test now but it would have to wait a few days/weeks so I can find several hours as I suspect it will take that much and I'm really short on time. I'll try to squeeze the work in asap.

Fair enough :) Let me know if you need help with the test, it shouldn't be too hard, you can start from one of the existing server tests, probably just add a test file with a straight horizontal line, use an inverted axis projection and it will fail with a vertical line in the DXF and it will be good with your patch.

**Comment by tudorbarascu** [comment]
@elpaso Test done, wondering if I could have done it shorter. Hope it's good, tested locally and it fails without patch, passes with patch. Thanks.

File: tests/src/python/test_qgsserver_wms_dxf.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            "LAYERS": "test_dxf_export",
+            "STYLES": ",",
+            "FORMAT": "application/dxf",
+            "SCALE": "500",
+            "FILE_NAME": "test_dxf_export.dxf"
+        }.items())])
+
+        r, h = self._result(self._execute_request(qs))
+
+        dxf = '/tmp/test_dxf_export.dxf'
```
Comment (elpaso) [inline tests/src/python/test_qgsserver_wms_dxf.py:51]: Please use QTemporaryDir or python's tempfile for temporary test output

File: tests/src/python/test_qgsserver_wms_dxf.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+from qgis.core import (QgsVectorLayer)
+
+# Needed on Qt 5 so that the serialization of XML is consistent among all executions
+os.environ['QT_HASH_SEED'] = '1'
+
+
+class PyQgsServerWMSGetMapDxf(QgsServerTestBase):
+
+    def test_dxf_export_works_with_reverse_axis_epsg(self):
+        qs = "?" + "&".join(["%s=%s" % i for i in list({
```
Comment (elpaso) [inline tests/src/python/test_qgsserver_wms_dxf.py:34]: Are you sure about this? I find it hard to follow but it looks like you are assigning the same value to both ends of `=`

File: tests/src/python/test_qgsserver_wms_dxf.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        dxf = '/tmp/test_dxf_export.dxf'
+        f = open(dxf, 'wb')
+        f.write(r)
+        f.close()
+
+        vl = QgsVectorLayer(dxf, "lyr", "ogr")
+        myMessage = ('Expected downloaded: %s dxf contains 1 line\nGot: %s\n' %
+                     (1, vl.featureCount()))
+
+        assert vl.featureCount() == 1, myMessage
```
Comment (elpaso) [inline tests/src/python/test_qgsserver_wms_dxf.py:60]: you can use self.assert... here too

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 4]
```diff
@@ -865,9 +865,55 @@ namespace QgsWms
       dxfLayers.append( QgsDxfExport::DxfLayer( vlayer, layerAttribute ) );
     }
 
+    //map extent
+    QgsRectangle mapExtent = mWmsParameters.bboxAsRectangle();
+
+    QString crs = mWmsParameters.crs();
+    if ( crs.compare( "CRS:84", Qt::CaseInsensitive ) == 0 )
```
Comment (elpaso) [inline src/server/services/wms/qgswmsrenderer.cpp:872]:
```suggestion
    if ( crs.compare( QStringLiteral( "CRS:84" ), Qt::CaseInsensitive ) == 0 )
```

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 5]
```diff
@@ -865,9 +865,55 @@ namespace QgsWms
...
+    {
+      crs = QString( "EPSG:4326" );
```
Comment (elpaso) [inline src/server/services/wms/qgswmsrenderer.cpp:874]:
```suggestion
      crs = QStringLiteral( "EPSG:4326" );
```

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 6]
```diff
@@ -865,9 +865,55 @@ namespace QgsWms
...
+      mapExtent.invert();
+    }
+    else if ( crs.isEmpty() )
+    {
+      crs = QString( "EPSG:4326" );
```
Comment (elpaso) [inline src/server/services/wms/qgswmsrenderer.cpp:879]:
```suggestion
      crs = QStringLiteral( "EPSG:4326" );
```

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 7]
```diff
@@ -865,9 +865,55 @@ namespace QgsWms
...
+    }
+
+    QgsCoordinateReferenceSystem outputCRS;
+
+    outputCRS = QgsCoordinateReferenceSystem::fromOgcWmsCrs( crs );
```
Comment (elpaso) [inline src/server/services/wms/qgswmsrenderer.cpp:884]: You can assign directly on line 882

**Review by elpaso — COMMENTED** [review]
LGTM, I just left some minor notes.

File: tests/src/python/test_qgsserver_wms_dxf.py (hunk 2 above)
Comment (tudorbarascu) [reply tests/src/python/test_qgsserver_wms_dxf.py:34]: I'm sure as it's in almost all the server tests (e.g. test_qgsserver_wms_getmap*). Should I do it differently?

File: tests/src/python/test_qgsserver_wms_dxf.py (hunk 1 above)
Comment (tudorbarascu) [reply tests/src/python/test_qgsserver_wms_dxf.py]: Done

File: tests/src/python/test_qgsserver_wms_dxf.py (hunk 3 above)
Comment (tudorbarascu) [reply tests/src/python/test_qgsserver_wms_dxf.py]: done

File: src/server/services/wms/qgswmsrenderer.cpp (hunk 4 above)
Comment (tudorbarascu) [reply src/server/services/wms/qgswmsrenderer.cpp]: done

**Comment by tudorbarascu** [comment]
@elpaso Thanks for the review. Everything addressed except the ``qs = "?" + "&".join(["%s=%s" % i for i in list({`` part which is the same as in all the other tests and it works. If it needs improvements I would need suggestions on how to proceed. Thanks again.

File: tests/src/python/test_qgsserver_wms_dxf.py (hunk 2 above)
Comment (elpaso) [reply tests/src/python/test_qgsserver_wms_dxf.py:36]: Nevermind, it's correct because `i` is a tuple.
